ompi/mca/hook/comm_method/hook_comm_method_fns.c

```diff
@@ -874,8 +873,375 @@ ompi_report_comm_methods(int called_from_location) // 1 = from init, 2 = from fi
+ *  return.
+ */
+static void
+sprint_bitmaps(char **affstring, int *affstring_cursz,
+    int indentation,
+    hwloc_cpuset_t *cpus_forrank, int nranks,
+    int *position)
+{
+    int i, r;
+    int nchildren_done[32], depth, some_cores_printed_under_containing_obj;
+    int nextchar;
+    hwloc_obj_t obj;
+    hwloc_bitmap_t cpus;
+    char *key = "abcdefghijklmnopqrstuvwxyzABCDEGHIJKLMNOPQRSTUVWXYZ";
+    int rankA, rankB;
+
+/* pre-computing how many ranks to use
+ */
+    cpus = hwloc_bitmap_alloc();
+    r = *position;
+    hwloc_bitmap_copy(cpus, cpus_forrank[r]);
+    while (r < nranks-1) {
+        if (!hwloc_bitmap_intersects(cpus, cpus_forrank[r+1]) &&
+            (r - *position + 1) < (int)strlen(key))
+        {
+            ++r;
+            hwloc_bitmap_or(cpus, cpus, cpus_forrank[r]);
+        } else {
+            break;
+        }
+    }
+    hwloc_bitmap_free(cpus);
+/* update position, save a rankA and rankB to note the range we'll use */
+    rankA = *position;
+    rankB = r;
+    *position = *position + (rankB - rankA + 1);
+
+    nextchar = strlen(*affstring);
+    LENGTHEN(affstring, nextchar, affstring_cursz, indentation + 1);
+    if (rankA != 0) {
+        (*affstring)[nextchar++] = '\n';
+        (*affstring)[nextchar] = 0;
+        for (i=0; i<indentation; ++i) {
+            (*affstring)[nextchar++] = ' ';
+        }
+        (*affstring)[nextchar] = 0;
+    }
+
+/* for bookkeeping:
+ * let the top of the stack be the obj we're processing, and keep
+ * nchildren_done in the stack for each obj
+ */
+    obj = hwloc_get_root_obj(opal_hwloc_topology);
+    depth = 0;
+    nchildren_done[depth] = 0;
+    while (obj) {
+        LENGTHEN(affstring, nextchar, affstring_cursz, 2);
+        if (obj->type == HWLOC_OBJ_PACKAGE) {
+            (*affstring)[nextchar++] = '[';
+            (*affstring)[nextchar] = 0;
+            some_cores_printed_under_containing_obj = 0;
+        }
+        if (obj->type == HWLOC_OBJ_CORE && some_cores_printed_under_containing_obj) {
+            (*affstring)[nextchar++] = '/';
+            (*affstring)[nextchar] = 0;
+        }
+        /* if this is a numa level it doesn't have .type NUMA, it just
+         * has a .memory_arity on the side.
+         * The below code prints the "<...>" for both interesting
+         * and non-interesting numas.  Eg if each package is a numa
+         * this might print "[<../..>]".  We could change this to only
+         * print the intersting cases by adding .type==GROUP here,
+         * because the GROUP .type is used when the numa level isn't a 1:1
+         * match for some other obj in the tree.
+         *
+         * The iszero() check is for skipping numas that have no PUs under
+         * them.  Without that check my PPC machines at least look like
+         * "[<../..>][<../..>]<><><><><><>"
+         */
+        if (obj->memory_arity > 0 && !hwloc_bitmap_iszero(obj->memory_first_child->cpuset)) {
```

Comment:
This line is not happy when compiled with `--no-oshmem`:

```
hook_comm_method_fns.c: In function 'sprint_bitmaps':
hook_comm_method_fns.c:1096:18: error: 'struct hwloc_obj' has no member named 'memory_arity'; did you mean 'memory'?
         if (obj->memory_arity > 0 && !hwloc_bitmap_iszero(obj->memory_first_child->cpuset)) {
                  ^~~~~~~~~~~~
                  memory
hook_comm_method_fns.c:1096:64: error: 'struct hwloc_obj' has no member named 'memory_first_child'; did you mean 'first_child'?
         if (obj->memory_arity > 0 && !hwloc_bitmap_iszero(obj->memory_first_child->cpuset)) {
                                                                ^~~~~~~~~~~~~~~~~~
                                                                first_child
hook_comm_method_fns.c:1122:22: error: 'struct hwloc_obj' has no member named 'memory_arity'; did you mean 'memory'?
             if (obj->memory_arity > 0 && !hwloc_bitmap_iszero(obj->memory_first_child->cpuset)) {
                      ^~~~~~~~~~~~
                      memory
hook_comm_method_fns.c:1122:68: error: 'struct hwloc_obj' has no member named 'memory_first_child'; did you mean 'first_child'?
             if (obj->memory_arity > 0 && !hwloc_bitmap_iszero(obj->memory_first_child->cpuset)) {
                                                                    ^~~~~~~~~~~~~~~~~~
                                                                    first_child
 ```

Reply:
Also, have you double-checked those definitions with HWLOC v1.11? I'm not sure those memory fields are present back there - best to check.

Reply:
That's a good point. If this is brought over to v5, that may need to get protected if it is not.

Reply:
(pushing another commit, intend to squash once confirmed it's what we want)

For this one I'd like to just go with "`#if HWLOC_API_VERSION >= 0x20000`" instead of trying to maintain a path that prints the numa level for the older hwloc.  The numa level is kind of nice, but I don't feel like it's so important as to maintain two paths for it.  So with the `#ifdef` I just added it will leave out the numa markers for older hwloc.